### DISCUSSION TITLE
Exploring New Horizons: Enhancing Hilla with Dynamic Code Generation

### DISCUSSION THREAD

**Discussion by cromoteca** [vaadin/hilla#1207] [Ideas]
In a recent internal meeting, we discussed some intriguing possibilities for introducing a powerful new feature that could help your development experience. We invite you to join us in this discussion and provide your valuable insights as we embark on this subject.

### The Vision

Our vision revolves around simplifying the development process by automating routine tasks and empowering developers with enhanced code generation capabilities. Here are some key points from our meeting that outline the direction we're considering:

- Create views based on available Spring singletons: Hilla endpoints, Spring services, and JPA repositories are all defined as Spring singletons. Hilla could automatically provide a starting point for any of those as CRUD views or forms.
- Dynamic Code Generation: all the steps from endpoint to view could be generated in a dynamic way, without polluting the application code.
- Ejection of Single Items: We plan to enable the ejection of single items along the ones described in the previous step, to allow developers to safely override and customize them.

### Your Input Matters

Now, here's where you come in. We want your feedback and insights on two critical questions:

- Auto-instantiation of Views vs. Immediate Code Generation: Do you prefer auto-instantiation of views that you can further customize, or should we generate code right away?
- JPA Repo vs. Service Layers: In your projects, do you typically rely more on JPA repositories or services? Understanding your preferences will help us fine-tune our approach to better suit your needs. Exposing a JPA repository is generally bad practice, but that can help prototyping.

And of course feel free to tell us about any other idea or suggestion on this topic.

**Comment by sveine** [comment]
JPA Repo vs. Service Layers

> I think Service Layers. I like clean JPA Repo, and use Service Layers

**Reply by tarekoraby** [reply]
@sveine, what could be your typical use case(s) for auto-generating code/view based on the service layer?

**Reply by sveinnetnordic** [reply]
Some loose thoughts : Annotate a service interface and have them all exposed front-end. For CRUD it could be something like @ClientCRUD(read = Service::getPeople, save = .., delete = ..., title = ...) or maybe point to a Repo.
